tensorflow/contrib/go/graph.go

```diff
@@ -0,0 +1,570 @@
+		def: &pb.NodeDef{
+			Name:   name,
+			Op:     opName,
+			Input:  inputs,
+			Device: device,
+			Attr:   make(map[string]*pb.AttrValue),
+		},
+		outDataTypes: make(map[string]DataType),
+	}
+
+	if attrs == nil {
+		attrs = make(map[string]interface{})
+	}
+	gr.matchTypes(input, node, attrs, op)
+
+	for _, attr := range op.Attr {
+		// Check if the attribute is specified, if it is not
+		// and doesn't have a default value, return an error since it
+		// is mandatory
+		if v, ok := attrs[attr.Name]; ok {
+			node.def.Attr[attr.Name] = gr.castAttrValue(attr.Type, v)
+			if node.def.Attr[attr.Name] == nil {
+				return nil, &ErrInvalidAttrValue{
+					operation:  opName,
+					attribName: attr.Name,
+				}
+			}
+		} else {
+			if attr.DefaultValue != nil {
+				node.def.Attr[attr.Name] = attr.DefaultValue
+			} else {
+				return nil, &ErrMandatoryAttributeNotSpecified{
+					operation:  opName,
+					attribName: attr.Name,
+				}
+			}
+		}
+	}
+
+	gr.def.Node = append(gr.def.Node, node.def)
+
+	return node, nil
+}
+
+// Variable Creates a variable operation and adds it to the graph. A variable
+// is a type of tensor that holds state in the form of a tensor that persists
+// across steps.
+func (gr *Graph) Variable(name string, initialData interface{}) (op *GraphNode, err error) {
+	var dims [][]int64
+
+	ts, err := NewTensor(initialData)
+	if err != nil {
+		return
+	}
+	gr.variables[name] = ts
+
+	shape := new(pb.TensorShapeProto)
+	if ts.NumDims() == 0 {
+		dims = [][]int64{{1}}
+	} else {
+		dims = ts.Shape()
+	}
+
+	shape.Dim = make([]*pb.TensorShapeProto_Dim, len(dims))
+	for i, dim := range dims {
+		shape.Dim[i] = &pb.TensorShapeProto_Dim{
+			Size: dim[i],
+		}
+	}
+
+	initVal, err := gr.Op("Const", name+"/initial_value", nil, "", map[string]interface{}{
+		"dtype": ts.DataType(),
+		"value": ts,
+		"shape": shape,
+	})
+	if err != nil {
+		return
+	}
+
+	variable, err := gr.Op("Variable", name, nil, "", map[string]interface{}{
+		"dtype":       ts.DataType(),
+		"shape":       shape,
+		"container":   "",
+		"shared_name": "",
+	})
+	if err != nil {
+		return
+	}
+
+	variable.ref = variable.def
+
+	_, err = gr.Op("Assign", name+"/Assign", []*GraphNode{variable, initVal}, "", map[string]interface{}{
+		"use_locking":    true,
+		"validate_shape": true,
+	})
+	if err != nil {
+		return
+	}
+
+	op, err = gr.Op("Identity", name+"/read", []*GraphNode{variable}, "", nil)
+
+	// For reference this variable, use the variable as it.
+	op.ref = variable.def
+
+	return
+}
+
+// String Returns a string representation of this graph, used for debugging
+// proposals.
+func (gr *Graph) String() string {
+	var bufStr bytes.Buffer
+	proto.MarshalText(&bufStr, gr.def)
+
+	return bufStr.String()
+}
+
+// addInitializationGraphOp Add the initialization operation to the graph to
+// cover all the added variables
+func (gr *Graph) addInitializationGraphOp() {
+	inputs := make([]string, len(gr.variables))
+	i := 0
+	for input := range gr.variables {
```

Comment:
no need to have i separate, range returns the index:
  for i, input := range gr.variables { ... }
(and then you don't need i++)